### PR TITLE
gnrc: create the basic "gnrc_mac" type for providing common MAC functionalities

### PR DESCRIPTION
**Pull request by zhuoshuguo**
The goal is to finally create/provide a "gnrc_mac" module which will provide common parameters and functionalities for MAC developers. 

This PR first introduces a basic "gnrc_mac_t" type which will be improved by latter PRs to contain more key parameters and functionalities that most MAC protocols will use/adopt, to increase code reuse for MAC  developers.

### CODE REVIEW

File: sys/include/net/gnrc/gnrc_mac/gnrc_mac_types.h [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+    TX_FEEDBACK_NOACK,
+    TX_FEEDBACK_BUSY
+} gnrc_mac_tx_feedback_t;
+
+/******************************************************************************/
+typedef struct gnrc_mac {
+    /* PID of MAC thread */
+    kernel_pid_t pid;
+    /* NETDEV device used by MAC */
+    gnrc_netdev2_t* netdev;
```
Comment (miri64) [inline sys/include/net/gnrc/gnrc_mac/gnrc_mac_types.h]: At least these to above are already provided by [`gnrc_netdev2_t`](https://github.com/RIOT-OS/RIOT/blob/master/sys/include/net/gnrc/netdev2.h#L75-L83), which should be used to implement MAC anyways by providing a `send` and `recv` implementation in my opinion. I also think that was how @kaspar030 intended it. Feel free to amend to that struct, but I think it could be beneficial to `#ifdef MODULE_GNRC_MAC` them out

File: sys/include/net/gnrc/gnrc_mac/gnrc_mac_types.h [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    const netdev2_driver_t* netdev2_driver;
```
Comment (miri64) [inline sys/include/net/gnrc/gnrc_mac/gnrc_mac_types.h]: This is incorporated in `gnrc_mac_t::netdev`. Why is this here?

File: sys/include/net/gnrc/gnrc_mac/gnrc_mac_types.h (hunk 1 above)
Comment (zhuoshuguo) [inline sys/include/net/gnrc/gnrc_mac/gnrc_mac_types.h]: Actually, this is inherited and extracted from the original [Lw-MAC](https://github.com/RIOT-OS/RIOT/pull/3730) module, and I didn't changed too much the original design.

But, anyway, that is a good idea to put some new MAC operations into gnrc_netdev2_t, I will see what I can do. :-)

**Comment by zhuoshuguo** [comment]
updated.

**Comment by miri64** [comment]
See https://github.com/zhuoshuguo/RIOT/pull/2

File: sys/include/net/gnrc/netdev2.h [hunk 3]
```diff
@@ -81,8 +96,44 @@ typedef struct gnrc_netdev2 {
      * @brief PID of this adapter for netapi messages
      */
     kernel_pid_t pid;
+
+#ifdef MODULE_GNRC_MAC
+    uint16_t mac_info;  /**< general information for the MAC protocol */
+#endif
 } gnrc_netdev2_t;
 
+bool gnrc_netdev2_get_rx_started(gnrc_netdev2_t *dev)
```
Comment (zhuoshuguo) [inline sys/include/net/gnrc/netdev2.h]: @miri64 should we `static inline` this function and others below here?

File: sys/include/net/gnrc/netdev2.h [hunk 4]
```diff
@@ -29,9 +29,13 @@
 #ifndef GNRC_NETDEV2_H
 #define GNRC_NETDEV2_H
 
+#include <assert.h>
+#include <stdint.h>
+
 #include "kernel_types.h"
 #include "net/netdev2.h"
 #include "net/gnrc.h"
+#include "net/gnrc/mac.h"
```
Comment (miri64) [inline sys/include/net/gnrc/netdev2.h]: This needs to be changed to `net/gnrc/mac/types.h`, too.

File: sys/include/net/gnrc/netdev2.h (hunk 3 above)
Comment (zhuoshuguo) [inline sys/include/net/gnrc/netdev2.h]: Yes, that was originally come from your code.
I will adapt them.  :-)

File: sys/include/net/gnrc/netdev2.h (hunk 4 above)
Comment (zhuoshuguo) [inline sys/include/net/gnrc/netdev2.h]: Yeah! Thanks.

File: sys/include/net/gnrc/netdev2.h [hunk 5]
```diff
@@ -81,8 +96,46 @@ typedef struct gnrc_netdev2 {
      * @brief PID of this adapter for netapi messages
      */
     kernel_pid_t pid;
+
+#ifdef MODULE_GNRC_MAC
+    uint16_t mac_info;  /**< general information for the MAC protocol */
+#endif
 } gnrc_netdev2_t;
 
+#ifdef MODULE_GNRC_MAC
```
Comment (zhuoshuguo) [inline sys/include/net/gnrc/netdev2.h]: @miri64 I also `#ifdef MODULE_GNRC_MAC` the `gnrc_mac` related functions, is it OK?

Comment (miri64) [inline sys/include/net/gnrc/netdev2.h]: Yes, helps to throw compiler errors if you use the functions if you not supposed to.

File: sys/include/net/gnrc/netdev2.h [hunk 6]
```diff
@@ -81,8 +96,81 @@ typedef struct gnrc_netdev2 {
...
+static inline bool gnrc_netdev2_get_rx_started(gnrc_netdev2_t *dev)
+{
+    return (dev->mac_info & GNRC_NETDEV2_MAC_INFO_RX_STARTED);
+}
+
+/**
+ * @brief set the rx_started state of the device
+ *
+ * This function is intended to be called only in the _event_cb function of
+ * the netdev2 device.
```
Comment (zhuoshuguo) [inline sys/include/net/gnrc/netdev2.h]: @miri64 could you help check the correctness of the description here? That is also how I understand the usage of these functions. Thanks

Comment (miri64) [inline sys/include/net/gnrc/netdev2.h]: Uhmm.. I don't know about the usage restrictions of this function... as is it can be called from anywhere in the MAC thread (but is restricted to that thread). @daniel-k was the rx_started boolean from your implementation supposed to only be called in the event callback?

Comment (daniel-k) [inline sys/include/net/gnrc/netdev2.h]: >  @daniel-k was the rx_started boolean from your implementation supposed to only be called in the event callback?

In my implementation `rx_started` was part of the global LwMAC state struct. I didn't follow the refactoring to that detail, so I don't know how introduced `gnrc_netdev2_set_rx_started()`, sorry :/

But this function might be dangerous in general. If I remember correctly, `_event_cb()` can be called in interrupt context as well as in thread context, but it is definetly not thread-safe. Furthermore, manipulating the rx_started information from outside the MAC layer will also cause malfunction.

Comment (miri64) [inline sys/include/net/gnrc/netdev2.h]: > In my implementation rx_started was part of the global LwMAC state struct. I didn't follow the refactoring to that detail, so I don't know who introduced gnrc_netdev2_set_rx_started(), sorry :/

That was me, but the functionality is the same as the `rx_started` variable, just a little bit more memory friendly ;-)

> But this function might be dangerous in general. If I remember correctly, _event_cb() can be called in interrupt context as well as in thread context, but it is definetly not thread-safe (docs should reflect that IMHO). Furthermore, manipulating the rx_started information from outside the MAC layer will also cause malfunction.

Okay then let's restrict the usage to the event callback. @zhuoshuguo please replace "`_event_cb` of the netdev2 device" with "`netdev2_t::event_callback()`" this way it gets properly referenced by the doc parser.

File: sys/include/net/gnrc/netdev2.h [hunk 7]
```diff
@@ -118,12 +137,28 @@ static inline void gnrc_netdev2_set_rx_started(gnrc_netdev2_t *dev, bool rx_star
...
 {
     return (gnrc_mac_tx_feedback_t)(dev->mac_info &
                                     GNRC_NETDEV2_MAC_INFO_TX_FEEDBACK_MASK);
 }
 
+/**
+ * @brief set the transmission feedback of the device
+ *
+ * This function is intended to be called only in the _event_cb function of
+ * the netdev2 device.
```
Comment (miri64) [inline sys/include/net/gnrc/netdev2.h]: Here, too.

**Comment by zhuoshuguo** [comment]
@miri64 with your agreement on the function description, updated. Sorry for the late reply. :-)

**Comment by miri64** [comment]
I'm fine with this PR now as is. Maybe we can get this into the release. Please squash.

**Comment by zhuoshuguo** [comment]
Squashed.

File: sys/include/net/gnrc/mac/types.h [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+#include <net/netdev2.h>
+#include <net/gnrc/netdev2.h>
+#include <net/gnrc.h>
+
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+typedef enum {
```
Comment (miri64) [inline sys/include/net/gnrc/mac/types.h]: This type (and its members) still requires documentation (please squash that immediately).

**Comment by miri64** [comment]
Postponed due to feature freeze.

**Comment by zhuoshuguo** [comment]
Updated, added documentation. @miri64 Sorry for the slow action.

**Comment by zhuoshuguo** [comment]
@miri64 Is this PR OK for merge now? :-)
